teps/0130-pipeline-level-build-cache.md

```diff
@@ -0,0 +1,557 @@
+  for example, if using `docker exec` or `docker-compose`
+
+Allowing multiple Tasks in a PipelineRun to use the same docker daemon allows PipelineRuns with
+docker builds to take advantage of Pipeline features (e.g. using matrix to build multiple images,
+executing a push conditionally on previous steps) while still having access to the same local
+cache of built images and avoid having to push to a registry in between.
+
+### Goals
+
+- More flexible build Pipelines
+- Build cache is only accessible to the single PipelineRun that uses it
+
+### Non-Goals
+
+- Sharing a daemon or build cache between PipelineRuns
+
+### Requirements
+
+- A docker image built during a Pipeline can be used in a subsequent Task without first pushing it to an image registry
+
+## Proposal
+
+Add support for Sidecars with the same lifespan as PipelineRuns, for example:
+
+```yaml
+kind: Pipeline
+spec:
+  params:
+  - name: image
+  - name: docker-host
+    value: $(sidecars.docker-daemon.host):2376
+  workspaces:
+  - name: source
+  - name: docker-tls-certs
+  results:
+  - name: image-digest
+    value: $(tasks.docker-build.results.IMAGE_DIGEST)
+  tasks:
+  - name: git-clone
+    ...
+  - name: docker-build
+    params:
+    - name: image
+    - name: docker-host
+    workspaces:
+    - name: source
+    taskSpec:
+      workingDir: $(workspaces.source.path)
+      steps:
+      - image: docker.io/library/docker
+        script: docker build -f ./Dockerfile -t $(params.image)
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+      results:
+      - name: IMAGE_DIGEST
+    runAfter: ["git-clone"]
+  # This task uses the built image without first pushing it to a registry
+  - name: pytest
+    params:
+    - name: docker-host
+    - name: image
+    taskSpec:
+      workingDir: $(workspaces.source.path)
+      steps:
+      - image: docker.io/library/docker
+        script: docker exec $(params.image) python -m pytest
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+    runAfter: ["docker-build"]
+  # This task pushes the image to the registry only if tests passed
+  - name: docker-push
+    when:
+    - input: "$(tasks.pytest.status)"
+      operator: in
+      values: ["Succeeded"]
+    params:
+    - name: image
+    - name: docker-host
+    taskSpec:
+      steps:
+      - image: docker.io/library/docker
+        script: docker push $(params.image)
+        env:
+        - name: DOCKER_HOST
+          value: $(params.docker-host)
+    runAfter: ["pytest"]
+  sidecars:
+  - name: docker-daemon
+    image: docker:dind
+    securityContext:
+      privileged: true
+    ports:
+    - 2376
+    startupProbe:
+      tcpSocket:
+        port: 2376
+    workspaces:
+    - name: docker-tls-certs   
+```
+
+Here's what the [docker-build catalog task](https://github.com/tektoncd/catalog/blob/81bf7dc5610d5fa17281940a72a6377604105cea/task/docker-build/0.1/docker-build.yaml)
```

Comment:
Note: this is what a `docker-build` `Task` could look like today as well. What is missing to make this "smarter" today, is a way to conditionnally start a sidecar (for example, in case of `params.docker_host` is empty or default value) and same for other possible requirements to run the sidecar (like `securityContext.privileged`, …)